jobs/schedulers.go

```diff
@@ -87,89 +98,95 @@ func (srv *JobServer) InitSchedulers() *Schedulers {
 	}
 
 	schedulers.nextRunTimes = make([]*time.Time, len(schedulers.schedulers))
-	return schedulers
+	srv.schedulers = schedulers
+
+	return nil
 }
 
-func (schedulers *Schedulers) Start() *Schedulers {
+// Start starts the schedulers. This call is not safe for concurrent use.
+// Synchronization should be implemented by the caller.
+func (schedulers *Schedulers) Start() {
 	schedulers.listenerId = schedulers.jobs.ConfigService.AddConfigListener(schedulers.handleConfigChange)
 
 	go func() {
-		schedulers.startOnce.Do(func() {
-			mlog.Info("Starting schedulers.")
-
-			defer func() {
-				mlog.Info("Schedulers stopped.")
-				close(schedulers.stopped)
-			}()
-
-			now := time.Now()
-			for idx, scheduler := range schedulers.schedulers {
-				if !scheduler.Enabled(schedulers.jobs.Config()) {
-					schedulers.nextRunTimes[idx] = nil
-				} else {
-					schedulers.setNextRunTime(schedulers.jobs.Config(), idx, now, false)
-				}
+		mlog.Info("Starting schedulers.")
+
+		defer func() {
+			mlog.Info("Schedulers stopped.")
+			close(schedulers.stopped)
+		}()
+
+		now := time.Now()
+		for idx, scheduler := range schedulers.schedulers {
+			if !scheduler.Enabled(schedulers.jobs.Config()) {
+				schedulers.nextRunTimes[idx] = nil
+			} else {
+				schedulers.setNextRunTime(schedulers.jobs.Config(), idx, now, false)
 			}
+		}
 
-			for {
-				timer := time.NewTimer(1 * time.Minute)
-				select {
-				case <-schedulers.stop:
-					mlog.Debug("Schedulers received stop signal.")
-					timer.Stop()
-					return
-				case now = <-timer.C:
-					cfg := schedulers.jobs.Config()
-
-					for idx, nextTime := range schedulers.nextRunTimes {
-						if nextTime == nil {
-							continue
-						}
+		for {
+			timer := time.NewTimer(1 * time.Minute)
+			select {
+			case <-schedulers.stop:
+				mlog.Debug("Schedulers received stop signal.")
+				timer.Stop()
+				return
+			case now = <-timer.C:
+				cfg := schedulers.jobs.Config()
 
-						if time.Now().After(*nextTime) {
-							scheduler := schedulers.schedulers[idx]
-							if scheduler != nil {
-								if schedulers.isLeader && scheduler.Enabled(cfg) {
-									if _, err := schedulers.scheduleJob(cfg, scheduler); err != nil {
-										mlog.Error("Failed to schedule job", mlog.String("scheduler", scheduler.Name()), mlog.Err(err))
-									} else {
-										schedulers.setNextRunTime(cfg, idx, now, true)
-									}
+				for idx, nextTime := range schedulers.nextRunTimes {
+					if nextTime == nil {
+						continue
+					}
+
+					if time.Now().After(*nextTime) {
+						scheduler := schedulers.schedulers[idx]
+						if scheduler != nil {
+							if schedulers.isLeader && scheduler.Enabled(cfg) {
```

Comment:
Would it make sense to trim this down to -> ` if scheduler := schedulers.schedulers[idx]; scheduler != nil && schedulers.isLeader && scheduler.Enabled(cfg) {`

Reply:
Hmm .. looks a bit too compressed for my taste :P 0/5.

Reply:
Didn't really get into that logic. If you prefer we can again use continue to remove some nesting:

```go
	if time.Now().After(*nextTime) {
		scheduler := schedulers.schedulers[idx]
		if scheduler == nil || !schedulers.isLeader || !scheduler.Enabled(cfg) {
			continue
		}
		if _, err := schedulers.scheduleJob(cfg, scheduler); err != nil {
			mlog.Error("Failed to schedule job", mlog.String("scheduler", scheduler.Name()), mlog.Err(err))
			continue
		}
		schedulers.setNextRunTime(cfg, idx, now, true)
	}
```

Reply:
☝️ this is even better 😂